### PR TITLE
Implement v3 main features

### PR DESCRIPTION
**Pull request by filipeesch**
# Description

This Pull Request implements the main features of 3.0 version:
- Improved Mechanism for Signalling Message Processing Completion #426
- Expose Worker Events to Client Applications #427
- Improve Dependency Injection Scope Management #428
- Dynamic Worker Pool Scaling #429

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [x] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: src/KafkaFlow/Consumers/Consumer.cs [hunk 1]
```diff
@@ -120,24 +120,29 @@ public IEnumerable<TopicPartitionLag> GetTopicPartitionsLag()
...
                 });
         }
 
-        public void Commit(IEnumerable<TopicPartitionOffset> offsetsValues)
+        public void Commit(IReadOnlyCollection<TopicPartitionOffset> offsets)
         {
-            this.consumer.Commit(offsetsValues);
+            var validOffsets = offsets
+                .Where(x => x.Offset.Value >= 0)
+                .ToList();
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/Consumer.cs:134]: A Span can be used here?

File: src/KafkaFlow/Consumers/ConsumerContext.cs [hunk 2]
```diff
@@ -2,50 +2,77 @@ namespace KafkaFlow.Consumers
...
 
         public bool ShouldStoreOffset { get; set; } = true;
 
-        public DateTime MessageTimestamp => this.kafkaResult.Message.Timestamp.UtcDateTime;
+        public DateTime MessageTimestamp { get; }
 
-        public void StoreOffset() => this.offsetManager.MarkAsProcessed(this.kafkaResult.TopicPartitionOffset);
+        public Task<TopicPartitionOffset> Completion => this.completionSource.Task;
+
+        public void StoreOffset()
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/ConsumerContext.cs:63]: Change the name to Complete()

File: src/KafkaFlow/Consumers/ConsumerManager.cs [hunk 3]
```diff
@@ -34,17 +42,69 @@ public Task StartAsync()
...
+
             await this.Feeder.StopAsync().ConfigureAwait(false);
             await this.WorkerPool.StopAsync().ConfigureAwait(false);
 
+            this.evaluateWorkersCountTimer?.Dispose();
+            this.evaluateWorkersCountTimer = null;
             this.Consumer.Dispose();
         }
 
+        public void Dispose()
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/ConsumerManager.cs:66]: remove this metthod?

File: src/KafkaFlow/Consumers/OffsetCommitter.cs [hunk 4]
```diff
@@ -5,45 +5,70 @@ namespace KafkaFlow.Consumers
...
                 pendingOffsetsHandlers,
             ILogHandler logHandler)
         {
             this.consumer = consumer;
+            this.resolver = resolver;
+            this.pendingOffsetsHandlers = pendingOffsetsHandlers;
             this.logHandler = logHandler;
+        }
+
+        public void FlushOffsets() => this.CommitHandler();
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/OffsetCommitter.cs:41]: is this method used?

File: src/KafkaFlow/Consumers/PartitionOffsets.cs [hunk 5]
```diff
@@ -3,47 +3,52 @@ namespace KafkaFlow.Consumers
...
-            lock (this.receivedOffsets)
+            lock (this.receivedContexts)
             {
-                this.receivedOffsets.AddLast(offset);
+                this.receivedContexts.AddLast(context);
             }
         }
 
-        public bool ShouldCommit(long offset, out long lastProcessedOffset)
+        public bool ShouldCommit(IConsumerContext context, out IConsumerContext lastProcessedContext)
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/PartitionOffsets.cs:21]: is IsEnableToCommit a better name?

File: src/KafkaFlow/Consumers/WorkerLifetimeContext.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+namespace KafkaFlow.Consumers
+{
+    internal class WorkerLifetimeContext : IWorkerLifetimeContext
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/WorkerLifetimeContext.cs:3]: try to think in a better name

File: samples/KafkaFlow.Sample.BatchOperations/PrintConsoleMiddleware.cs [hunk 7]
```diff
@@ -5,18 +5,47 @@
...
+    }
+
+    public async Task OnNotification()
+    {
+        this.buffer.Writer.Complete();
+        await this.task;
     }
-}
\ No newline at end of file
+}
```
Comment (filipeesch) [inline samples/KafkaFlow.Sample.BatchOperations/PrintConsoleMiddleware.cs:51]: undo the changes in this file

File: samples/KafkaFlow.Sample.BatchOperations/Program.cs [hunk 8]
```diff
@@ -53,30 +59,41 @@
...
-            await bus.StopAsync();
-            return;
+    if (input!.Equals("exit", StringComparison.OrdinalIgnoreCase))
+    {
+        await bus.StopAsync();
+        break;
     }
-}
\ No newline at end of file
+}
```
Comment (filipeesch) [inline samples/KafkaFlow.Sample.BatchOperations/Program.cs:99]: undo the changes in this file

File: src/KafkaFlow.Abstractions/Configuration/WorkersCountContext.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        public string ConsumerName { get; }
+
+        public string ConsumerGroupId { get; }
+
+        /// <summary>
+        /// Gets the assigned partitions to the consumer
+        /// </summary>
+        public IReadOnlyCollection<TopicPartitions> AssignedTopicsPartitions { get; }
+    }
+}
```
Comment (filipeesch) [inline src/KafkaFlow.Abstractions/Configuration/WorkersCountContext.cs:29]: add XML comments in this file

File: src/KafkaFlow/Consumers/OffsetCommitter.cs (hunk 4 above)
Comment (filipeesch) [reply src/KafkaFlow/Consumers/OffsetCommitter.cs:41]: removed

File: src/KafkaFlow/Consumers/ConsumerManager.cs (hunk 3 above)
Comment (filipeesch) [reply src/KafkaFlow/Consumers/ConsumerManager.cs:66]: removed

File: src/KafkaFlow/Consumers/ConsumerContext.cs (hunk 2 above)
Comment (filipeesch) [reply src/KafkaFlow/Consumers/ConsumerContext.cs:63]: Changed. Also changed the Configuration method from `WithManualStoreOffsets` to `WithManualMessageCompletion` and the `ConsumerContext` property from `ShouldStoreOffset` to `AutoCompleteMessage`

File: src/KafkaFlow/Consumers/Consumer.cs (hunk 1 above)
Comment (filipeesch) [reply src/KafkaFlow/Consumers/Consumer.cs:134]: no, it isn't possible

File: src/KafkaFlow.BatchConsume/BatchConsumeMiddleware.cs [hunk 10]
```diff
@@ -5,19 +5,28 @@
...
         private readonly TimeSpan batchTimeout;
         private readonly ILogHandler logHandler;
+        private readonly IConsumerConfiguration consumerConfiguration;
 
         private readonly List<IMessageContext> batch;
         private CancellationTokenSource dispatchTokenSource;
+        private Task<Task> dispatchTask;
 
         public BatchConsumeMiddleware(
+            IConsumerMiddlewareContext workerContext,
```
Comment (brmagadutra) [inline src/KafkaFlow.BatchConsume/BatchConsumeMiddleware.cs:29]: This isn't only related to the worker context, right?

File: src/KafkaFlow.BatchConsume/KafkaFlow.BatchConsume.csproj [hunk 11]
```diff
@@ -11,4 +11,8 @@
       <ProjectReference Include="..\KafkaFlow\KafkaFlow.csproj" />
     </ItemGroup>
 
+    <ItemGroup>
+      <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
```
Comment (simaoribeiro) [inline src/KafkaFlow.BatchConsume/KafkaFlow.BatchConsume.csproj:15]: This reference is not necessary

**Review by simaoribeiro — APPROVED** [review]

**Review by ailtonguitar — COMMENTED** [review]
Why not to remove the VoidObject and use an interface like ISubject<WorkerStoppingSubject> if there are no arguments to notify?

**Comment by simaoribeiro** [comment]
> Why not to remove the VoidObject and use an interface like ISubject if there are no arguments to notify?

I agree with @ailtonguitar . The VoidObject usage looks akward

File: src/KafkaFlow/Consumers/WorkersBalancers/ConsumerLagWorkerBalancer.cs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    internal class ConsumerLagWorkerBalancer
+    {
+        private readonly IClusterManager clusterManager;
+        private readonly IConsumerAccessor consumerAccessor;
+        private readonly ILogHandler logHandler;
+        private readonly int totalConsumerWorkers;
+        private readonly int minInstanceWorkers;
+        private readonly int maxInstanceWorkers;
+
+        public ConsumerLagWorkerBalancer(
```
Comment (ailtonguitar) [inline src/KafkaFlow/Consumers/WorkersBalancers/ConsumerLagWorkerBalancer.cs:19]: Is it possible to provide some documentation here about the math used to find the number of workers based on the lags?

File: src/KafkaFlow/DateTimeProvider.cs [hunk 13]
```diff
@@ -6,6 +6,8 @@ internal class DateTimeProvider : IDateTimeProvider
     {
         public DateTime Now => DateTime.Now;
 
+        public DateTime UtcNow => DateTime.UtcNow;
```
Comment (ailtonguitar) [inline src/KafkaFlow/DateTimeProvider.cs:9]: Is the "Now" property still needed?

File: src/KafkaFlow.Abstractions/IConsumerContext.cs [hunk 14]
```diff
@@ -49,14 +55,43 @@ public interface IConsumerContext
...
+        /// that are scoped to the lifecycle of a single consumer.
+        /// </summary>
+        IDependencyResolver ConsumerDependencyResolver { get; }
+
+        /// <summary>
+        /// Gets an instance of IDependencyResolver which provides methods to resolve dependencies.
+        /// This instance is tied to the worker scope, meaning it is capable of resolving dependencies
+        /// that are scoped to the lifecycle of a single worker.
+        /// </summary>
+        IDependencyResolver WorkerDependencyResolver { get; }
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/IConsumerContext.cs:79]: Because each container instance has a different lifetime, their scopes are disposed in different moments. For the Microsoft DI consumer's instances or worker's instances both are "scoped", KafkaFlow manually manages these scopes to create custom lifetimes, such as consumer lifetime and worker lifetime.

File: src/KafkaFlow.Abstractions/VoidObject.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
+namespace KafkaFlow;
+
+/// <summary>
+/// A type that represents an empty object that should be ignored
+/// </summary>
+public class VoidObject
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/VoidObject.cs:6]: To create an ISubject with no parameter we need to duplicate all the observer implementation one with a parameter and the other with no parameter. The alternative is to create a dummy type (`VoidObject`) and have only one implementation. Which do you prefer?

File: src/KafkaFlow.BatchConsume/BatchConsumeMessageContext.cs [hunk 16]
```diff
@@ -3,14 +3,17 @@ namespace KafkaFlow.BatchConsume
...
     {
+        private readonly IDependencyResolverScope batchDependencyScope;
+
         public BatchConsumeMessageContext(
             IConsumerContext consumer,
             IReadOnlyCollection<IMessageContext> batchMessage)
         {
             this.ConsumerContext = consumer;
             this.Message = new Message(null, batchMessage);
+            this.batchDependencyScope = consumer.WorkerDependencyResolver.CreateScope();
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.BatchConsume/BatchConsumeMessageContext.cs:16]: A Scope is a lightweight object. IMHO, creating a lazy object here will make the code more complex with practically no performance benefits.

File: src/KafkaFlow/Consumers/WorkersBalancers/ConsumerLagWorkerBalancer.cs (hunk 12 above)
Comment (filipeesch) [reply src/KafkaFlow/Consumers/WorkersBalancers/ConsumerLagWorkerBalancer.cs:24]: Added summary comment in the class explaining the used math

File: src/KafkaFlow/DateTimeProvider.cs (hunk 13 above)
Comment (filipeesch) [reply src/KafkaFlow/DateTimeProvider.cs:7]: removed

File: src/KafkaFlow.BatchConsume/KafkaFlow.BatchConsume.csproj (hunk 11 above)
Comment (filipeesch) [reply src/KafkaFlow.BatchConsume/KafkaFlow.BatchConsume.csproj]: removed

File: src/KafkaFlow.BatchConsume/BatchConsumeMiddleware.cs (hunk 10 above)
Comment (filipeesch) [reply src/KafkaFlow.BatchConsume/BatchConsumeMiddleware.cs]: renamed. :eyes:

**Review by joelfoliveira — APPROVED** [review]

File: src/KafkaFlow.Compressor.Gzip/GzipMessageCompressor.cs [hunk 17]
```diff
@@ -7,7 +7,6 @@
     /// <summary>
     /// A GZIP message compressor
     /// </summary>
-    [Obsolete("Compressors should only be used in backward compatibility scenarios, in the vast majority of cases native compression (producer.WithCompression()) should be used instead")]
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Compressor.Gzip/GzipMessageCompressor.cs:10]: Although this was previously marked as obsolete, it has become clear that removing this implementation isn't possible. Many users rely on this middleware for various purposes, and the obsolete tag has led to significant confusion. I think we should retain this implementation to ensure continued support for those users.

File: src/KafkaFlow.BatchConsume/BatchConsumeMiddleware.cs [hunk 18]
```diff
@@ -26,6 +35,9 @@ internal class BatchConsumeMiddleware : IMessageMiddleware, IDisposable
             this.batchTimeout = batchTimeout;
             this.logHandler = logHandler;
             this.batch = new(batchSize);
+            this.consumerConfiguration = workerContext.Consumer.Configuration;
+
+            workerContext.Worker.WorkerStopped.Subscribe(this);
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.BatchConsume/BatchConsumeMiddleware.cs]: We should utilize the stopped to ensure all buffered messages within the worker are delivered to the application

File: src/KafkaFlow.Abstractions/IConsumerContext.cs [hunk 19]
```diff
@@ -49,14 +55,43 @@ public interface IConsumerContext
         DateTime MessageTimestamp { get; }
 
         /// <summary>
-        /// Gets or sets a value indicating whether if the framework should store the current offset in the end when auto store offset is used
+        /// Gets or sets a value indicating whether if the framework should auto complete the message in the end
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/IConsumerContext.cs]: done

File: src/KafkaFlow.Abstractions/IConsumerContext.cs [hunk 20]
```diff
@@ -49,14 +55,43 @@ public interface IConsumerContext
...
+        /// </summary>
+        bool AutoMessageCompletion { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether if the message offset must be stored when the message is marked as completed
+        /// </summary>
+        bool StoreOffsetOnCompletion { get; set; }
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/IConsumerContext.cs]: agreed

File: src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs [hunk 21]
```diff
@@ -91,6 +92,23 @@ public interface IConsumerConfigurationBuilder
...
+        /// </summary>
+        /// <param name="calculator">A function that takes a WorkersCountContext object and returns a Task yielding the new workers count</param>
+        /// <param name="evaluationInterval">The interval that the calculator will be called</param>
+        /// <returns>The IConsumerConfigurationBuilder instance for method chaining</returns>
+        IConsumerConfigurationBuilder WithWorkersCount(
+            Func<WorkersCountContext, IDependencyResolver, Task<int>> calculator,
+            TimeSpan evaluationInterval);
+
+        /// <summary>
+        /// Configures a custom function to dynamically calculate the number of workers.
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs:106]: done

File: src/KafkaFlow.Abstractions/VoidObject.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ailtonguitar) [reply src/KafkaFlow.Abstractions/VoidObject.cs:6]: IMO it is better to think first in avoid to expose unnecessary things to outside of the framework (in this case VoidObject), regarding the internal observer implementation it is possible to refactor or to think about an alternative to not duplicate the whole logic without the need of breaking changes.

**Review by brmagadutra — APPROVED** [review]

File: src/KafkaFlow.Abstractions/Observer/Subject.cs [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+        /// </summary>
+        /// <param name="arg">The parameter passed by the client.</param>
+        /// <returns>A task representing the asynchronous notification operation.</returns>
+        public async Task NotifyAsync(TArg arg)
+        {
+            foreach (var observer in this.observers)
+            {
+                try
+                {
+                    await observer.OnNotification((TSubject)this, arg);
```
Comment (joelfoliveira) [inline src/KafkaFlow.Abstractions/Observer/Subject.cs:44]: A object that subscribes notifications that is disposed could lead to a NullReferenceException, correct?

Should this behavior result in a log error? we should probably do a null check instead for this use case, no?

Comment (filipeesch) [reply src/KafkaFlow.Abstractions/Observer/Subject.cs:44]: I like the unsubscribe method approach, it could be helpful in other scenarios

File: src/KafkaFlow.Abstractions/IDateTimeProvider.cs [hunk 23]
```diff
@@ -7,8 +7,8 @@ namespace KafkaFlow
     /// </summary>
     public interface IDateTimeProvider
     {
-        /// <inheritdoc cref="DateTime.Now"/>
-        DateTime Now { get; }
+        /// <inheritdoc cref="DateTime.UtcNow"/>
+        DateTime UtcNow { get; }
```
Comment (gsferreira) [inline src/KafkaFlow.Abstractions/IDateTimeProvider.cs:11]: Since we are changing, what about moving to use `DateTimeOffset`?
 
https://learn.microsoft.com/en-us/dotnet/api/system.timeprovider.getutcnow?view=net-8.0

File: src/KafkaFlow.Abstractions/IProducerContext.cs [hunk 24]
```diff
@@ -19,5 +19,12 @@ public interface IProducerContext
         /// Gets the partition offset associated with the message
         /// </summary>
         long? Offset { get; }
+
+        /// <summary>
+        /// Gets an instance of IDependencyResolver which provides methods to resolve dependencies.
+        /// This instance is tied to the producer scope, meaning it is capable of resolving dependencies
+        /// that are scoped to the lifecycle of a single producer.
+        /// </summary>
+        IDependencyResolver ProducerDependencyResolver { get; }
```
Comment (gsferreira) [inline src/KafkaFlow.Abstractions/IProducerContext.cs:28]: Do we need the Producer prefix here? 
Asking, since I'm looking at the IMessageContext at the same time.

File: src/KafkaFlow.UnitTests/OffsetCommitterTests.cs [hunk 25]
```diff
@@ -4,6 +4,7 @@ namespace KafkaFlow.UnitTests
     using System.Collections.Generic;
     using System.Linq;
     using System.Threading;
+    using System.Threading.Tasks;
     using Confluent.Kafka;
     using FluentAssertions;
```
Comment (JoaoRodriguesGithub) [inline src/KafkaFlow.UnitTests/OffsetCommitterTests.cs:9]: Please consider removing unnecessary usings. Also [this one](https://github.com/Farfetch/kafkaflow/pull/409/files#diff-20a8b61d7d160af60b910dce35d003fbdbcc7242cb908fe3aa6c26bc53cd5823R6)

**Review by erik-catalao — APPROVED** [review]
LGTM

**Comment by joelfoliveira** [comment]
Besides the migration guide that will be done afterward, should the documentation for dynamic workers be included in this PR?

Maybe update this documentations website/docs/guides/consumers/consumers.md#workers

File: samples/KafkaFlow.Sample.PauseConsumerOnError/PauseConsumerOnExceptionMiddleware.cs [hunk 26]
```diff
@@ -21,7 +21,7 @@ public async Task Invoke(IMessageContext context, MiddlewareDelegate next)
         }
         catch (Exception exception)
         {
-            context.ConsumerContext.ShouldStoreOffset = false;
+            context.ConsumerContext.AutoMessageCompletion = false;
```
Comment (joelfoliveira) [inline samples/KafkaFlow.Sample.PauseConsumerOnError/PauseConsumerOnExceptionMiddleware.cs:24]: Besides the migration guide that will be done afterward, should the documentation for dynamic workers be included in this PR?

Maybe update this documentations website/docs/guides/consumers/consumers.md#workers

**Review by joelfoliveira — APPROVED** [review]

File: samples/KafkaFlow.Sample.PauseConsumerOnError/PauseConsumerOnExceptionMiddleware.cs (hunk 26 above)
Comment (filipeesch) [reply samples/KafkaFlow.Sample.PauseConsumerOnError/PauseConsumerOnExceptionMiddleware.cs:24]: I have the documentation implemented in another branch. I'll open a new PR with it.

File: src/KafkaFlow.Abstractions/Consumers/IWorker.cs [hunk 27]
```diff
@@ -17,5 +19,15 @@ public interface IWorker
         /// </summary>
         /// <param name="handler"><see cref="Action"/> to be executed</param>
         void OnTaskCompleted(Action handler);
```
Comment (joelfoliveira) [inline src/KafkaFlow.Abstractions/Consumers/IWorker.cs:21]: As discussed in the PR #438 ([this discussion](https://github.com/Farfetch/kafkaflow/pull/438#discussion_r1341949396)), should we make this an event also?

File: src/KafkaFlow.Abstractions/Observer/Subject.cs (hunk 22 above)
Comment (joelfoliveira) [reply src/KafkaFlow.Abstractions/Observer/Subject.cs:44]: I added the same suggestion in @ailtonguitar PR #438 - [(discussion here)](https://github.com/Farfetch/kafkaflow/pull/438#discussion_r1341956898)

File: src/KafkaFlow.Abstractions/IProducerContext.cs (hunk 24 above)
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/IProducerContext.cs:28]: agreed

File: src/KafkaFlow/Consumers/WorkersBalancers/ConsumerLagWorkerBalancer.cs [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+                .ToList();
+        }
+
+        private async Task<int> CalculateAsync(WorkersCountContext context)
+        {
+            try
+            {
+                if (!context.AssignedTopicsPartitions.Any())
+                {
+                    return 1;
```
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow/Consumers/WorkersBalancers/ConsumerLagWorkerBalancer.cs:95]: done

File: src/KafkaFlow.UnitTests/OffsetCommitterTests.cs (hunk 25 above)
Comment (filipeesch) [reply src/KafkaFlow.UnitTests/OffsetCommitterTests.cs:9]: done

File: src/KafkaFlow.Abstractions/IDateTimeProvider.cs (hunk 23 above)
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/IDateTimeProvider.cs:11]: It appears that the TimeProvider can only be accessed in .NET 8.

**Review by ruiqbarbosa — APPROVED** [review]

**Review by JoaoRodriguesGithub — APPROVED** [review]
